src/toil/bus.py

```diff
@@ -599,5 +623,84 @@ def __init__(self) -> None:
+
+        We must not have connected to anything yet.
+        """
+
+        # We need to call the two inherited connection methods
+        super()._set_bus_and_message_types(bus, wanted_types)
+        self._set_bus(bus)
 
 
+def replay_message_bus(path: str) -> Dict[str, Any]:
+    """
+    Replay all the messages and work out what they mean for jobs.
+
+    We track the state and name of jobs here, by ID.
+    We would use a list of two items but MyPy can't understand a list
+    of items of multiple types, so we need to define a new class.
+    """
+    @dataclass
+    class JobStatus:
+        """
+        Records the status of a job.
+        """
+
+        job_store_id: str
+        name: str
+        exit_code: int
+        annotations: Dict[str, str]
+        toil_batch_id: int
+        external_batch_id: str
+        batch_system: str
+
+        def toJSON(self) -> str:
+            return json.dumps(self, default= lambda o: o.__dict__, indent=4)
+
+    job_statuses: Dict[str, JobStatus] = collections.defaultdict(lambda: JobStatus('', '', -1, {}, -1, '', ''))
+    batch_to_job_id = {}
+    with open(path, 'rb') as log_stream:
+        # Read all the full, properly-terminated messages about job updates
+        for event in MessageBus.scan_bus_messages(log_stream, [JobUpdatedMessage, JobIssuedMessage, JobCompletedMessage,
+                                                               JobFailedMessage, JobAnnotationMessage, BatchSystemMessage]):
+            # And for each of them
+            logger.info('Got message from workflow: %s', event)
+
+            if isinstance(event, JobUpdatedMessage):
+                # Apply the latest return code from the job with this ID.
+                job_statuses[event.job_id].exit_code = event.result_status
+            elif isinstance(event, JobIssuedMessage):
+                job_statuses[event.job_id].job_store_id = event.job_id
+                job_statuses[event.job_id].name = event.job_type
+                job_statuses[event.job_id].toil_batch_id = event.toil_batch_id
+                job_statuses[event.job_id].exit_code = -1
+                batch_to_job_id[event.toil_batch_id] = event.job_id
+            elif isinstance(event, JobCompletedMessage):
+                job_statuses[event.job_id].name = event.job_type
+                job_statuses[event.job_id].exit_code = event.exit_code
+            elif isinstance(event, JobFailedMessage):
+                job_statuses[event.job_id].name = event.job_type
+                if job_statuses[event.job_id].exit_code == 0:
+                    # Record the failure if we never got a failed exit code.
+                    job_statuses[event.job_id].exit_code = 1
+            elif isinstance(event, JobAnnotationMessage):
+                # Remember the last value of any annotation that is set
+                job_statuses[event.job_id].annotations[event.annotation_name] = event.annotation_value
```

Comment:
Since nobody ever sends a `JobAnnotationMessage` anymore, and instead we always send `BatchSystemMessage` to communicate e.g. the AWS Batch job ID, then the annotations here will always be empty.

1. If they are always empty, it might make sense to remove the field, and the `JobAnnotationMessage` altogether.
2. Whether you do that or not, properly supporting AGC's dialect of WES requires always including the AWS Batch job ID in the task names that the server reports. Right now we do that by [selecting a filter function](https://github.com/DataBiosphere/toil/blob/88d4589592eba433d10b3c835ef2a0be5653336a/src/toil/server/wes/toil_backend.py#L602-L611) that gets to see the annotations and modify the WES tasks, and then [applying the filter function](https://github.com/DataBiosphere/toil/pull/4292/files#diff-e05a9b10abd96d5732ff2a3577a21843f0c6efa1074f2e44d752670449124edbR238-R249) when we go to generate the tasks. But if we no longer keep the AWS Batch job ID in the annotations, the current filter function for AWS Batch won't work, and we will break AWS Batch support.

We really ought to have the AWS WES dialect under test if we want to not break it. But, failing that, we definitely need to adapt the filter function setup to be able to put the new `external_batch_id` field from the `JobStatus` into the task name, instead of looking at the annotations. Maybe we need to give the filter_function access to the whole `JobStatus` and not just the annotations field?

Reply:
This should probably be its own issue